cpp/src/transpose/transpose.cu

```diff
@@ -0,0 +1,191 @@
+/* Copyright 2018 NVIDIA Corporation.  All rights reserved. */
+
+#include "utilities/nvtx/nvtx_utils.h"
+#include "utilities/type_dispatcher.hpp"
+#include "rmm/thrust_rmm_allocator.h"
+#include <cudf.h>
+#include <cub/cub.cuh>
+#include <memory>
+#include <stdio.h>
+
+constexpr int WARP_SIZE = 32;
+constexpr int MAX_GRID_SIZE = (1<<16)-1;
+
+template <typename ColumnType>
+__global__
+void gpu_transpose(ColumnType **in_cols, ColumnType **out_cols,
+                  gdf_size_type ncols, gdf_size_type nrows)
+{
+  gdf_size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  gdf_size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+    
+  gdf_size_type stride_x = blockDim.x * gridDim.x;
+  gdf_size_type stride_y = blockDim.y * gridDim.y;
+
+  for(gdf_size_type i = x; i < ncols; i += stride_x)
+  {
+    for(gdf_size_type j = y; j < nrows; j += stride_y)
+    {
+      out_cols[j][i] = in_cols[i][j];
+    }
+  }
+}
+
+__global__
+void gpu_transpose_valids(gdf_valid_type **in_cols_valid,
+                          gdf_valid_type **out_cols_valid,
+                          gdf_size_type *out_cols_null_count,
+                          gdf_size_type ncols, gdf_size_type nrows)
+{
+  using MaskType = uint32_t;
+  constexpr uint32_t BITS_PER_MASK{sizeof(MaskType) * 8};
+
+  gdf_size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  gdf_size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  gdf_size_type stride_x = blockDim.x * gridDim.x;
+  gdf_size_type stride_y = blockDim.y * gridDim.y;
+
+  MaskType* const __restrict__ out_mask32 =
+    reinterpret_cast<MaskType*>(out_cols_valid[y]);
+
+  for(gdf_size_type i = x; i < ncols; i += stride_x)
+  {
+    for(gdf_size_type j = y; j < nrows; j += stride_y)
+    {
+      auto active_threads = __ballot_sync(0xffffffff, i < ncols && j < nrows);
```

Comment:
[The manual](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-vote-functions) says `__ballot_sync` will

> return an integer whose Nth bit is set if and only if predicate evaluates to non-zero for the Nth thread of the warp and the Nth thread is active.

So we probably don't need to calculate active threads mask from now on.

Reply:
So here's the relevant line from the manual as to why your code isn't safe.

>a mask must be passed that specifies the threads participating in the call. A bit, representing the thread's lane ID, must be set for each participating thread to ensure they are properly converged before the intrinsic is executed by the hardware. **All active threads named in mask must execute the same intrinsic with the same mask**, or the result is undefined.

You're passing in `0xffffffff` as the mask of active threads, telling it that all of the threads in the warp will execute the same `__ballot_sync` intrinsic. However, this isn't true, because for some warps, not all of the threads will participate. 

This is why you need to compute the active mask of threads _before_ the `for` loop, and then use that computed mask inside the `for` loop.

```  
auto active_threads = __ballot_sync(0xffffffff, i < ncols && j < nrows);
for(gdf_size_type i = x; i < ncols; i += stride_x)
  {
    for(gdf_size_type j = y; j < nrows; j += stride_y)
    {
      active_threads = __ballot_sync(active_threads, i < ncols && j < nrows);
```

Reply:
In fact, I believe using the `for` loop here isn't safe either, and a `while` loop would be better, because you need to recompute `active_threads` after the iterators have incremented.


```
auto active_threads = __ballot_sync(0xffffffff, i < ncols && j < nrows);
while( i < ncols)
{
    while(j < nrows)
    {
      ....
      j += stride_y;  
      active_threads = __ballot_sync(active_threads, i < ncols && j < nrows);
    }
    i += stride_x;
    active_threads = __ballot_sync(active_threads, i < ncols && j < nrows);
}
```

The `while` loop approach is necessary because you need to recompute `active_threads` after the iterators are incremented by `stride`. 

However, in the `for` loop approach, those threads that evaluate `i < ncols && j < nrows` as false for the next iteration will **not** participate in the computation of `active_threads` because they will have already dropped out of the `for` loop.

This is why the `while` loop approach is necessary such that you can update the iterators by `stride` and _then_ recompute the next iterations `active_threads`.

Reply:
Jake please resolve this if you are happy.